### ISSUE TITLE
Race condition in database creation?

### ISSUE THREAD

**Issue by sfayer** [DIRACGrid/DIRAC#4579]
Hi,

As @marianne013 mentioned on the mailing list, we've been having trouble installing a DIRAC server from scratch with dirac-setup-site. We put a lot more debugging around the error messages and we found the problem was multiple components trying to create the same tables at about the same time (the ones that make tables at start-up rather than installing an SQL file). One of them will make some of the tables and the others will get stuck with a mysterious message about foreign keys (explained below). I'm not sure why no-one has seen this before or why the winner of the race doesn't just make all of the tables successfully.  We've seen this with the TaskQueueDB and SystemLoggingDB, but think it could probably happen with any database used by more than one service. We added a long wait between each component being started and this appears to have mostly fixed our problem (we still got one failure, but perhaps our wait just wasn't long enough: Our test node is pretty slow).

Does anyone have any ideas about this? Is there any improvements we can add to the code to make the database creation more deterministic? Perhaps we can add some kind of DB lock around the table creation?

The foreign key error is actually a secondary bug. The _createTables function in the MySQL utilities module is generally only given a list of tables that don't exist by the caller. It then tries to order the tables so that the foreign keys are ordered correctly, but if there are pre-existing tables it doesn't take these into account. It reaches the conclusion that it can't find an order that would work and prints the misleading error message "Recursive Foreign Keys". Perhaps it would be simpler to always give it the tables in the correct creation order and remove this complex bit of ordering code or something like that? (Or make it so that it can cope with tables that already exist and then always give it the full table list?).

Regards,
Simon

**Comment by fstagni** [comment]
Hi,
one more info: once installing a DIRAC (or extendedDIRAC) from scratch, do you
1) want several components (i.e. services in this case) to be started at the same time (e.g. 3 JobManager services, who would all access TaskQueueDB)? 
2) just want to run one component per type?

Option 1 was never done before: in GithubActions for example we start only one service per type, and we don't see the issue you mention.

My suggestion would be anyway to look for fixes in the `ComponentInstaller.setupSite`.

**Comment by marianne013** [comment]
Hi Federico, We were trying to install a very basic version of DIRAC, i.e. only one Component of each type. Unfortunately I seem to have overwritten the modified files, but if I remember correctly all we did in the end was to increase a sleep in 
FrameworkSystem/Client/ComponentInstaller.py

**Comment by fstagni** [comment]
What you are doing is exactly what's being done at every commit in GitHub Actions, and there we've never seen the problems you mention. At the same time there is, for what I remember right now, only one database accessed by more than one service, and this is TaskQueueDB. If there's a slowness in creation of the tables, it's due to the MySQL (or MariaDB) instance. Can you try with this fix? https://github.com/DIRACGrid/DIRAC/compare/rel-v7r0...fstagni:v7r0-fixes59?expand=1

**Comment by sfayer** [comment]
Hi,

We set-up a new VM to test this and did another install but weren't able to reproduce our original issue: I suspect it only happens if the disk behind mysql is particularly slow.

We tried your fix out and the install was also ran through fine. My suggestion is that the fix is included in the hope that it does resolve the issue and we'll do a lot more debugging if we ever see the issue again.

Regards,
Simon
